docs/AsyncWrap/README.md

```diff
@@ -1,15 +1,276 @@
 Node.js tracing - AsyncWrap
 ================================================================================
 
-AsyncWrap is the central location in Node.js that handles async things
-specific to Node.js.  As such, it's a likely place where tracing will be
-desired.  Should serve as a good customer of the more general tracing story for
-Node.js.
+AsyncWrap is two things. One is a
+[class abstraction](https://github.com/nodejs/node/blob/master/src/async-wrap.h)
+there provides node internal facilities for handling async things, such as
+calling a callback.
+The other part is a an API for setting up hooks, that allow one to get
+structural tracing information about the life of handle objects. In the context
+of tracing the latter is usually what is meant.
 
-An older-ish intro to AsyncWrap is available in Trevor Norris's blog post
-[AsyncWrap Tutorial - An Introduction](http://blog.trevnorris.com/2015/02/asyncwrap-tutorial-introduction.html).
+_The reasoning for the current naming confusion, is that the API part implements
+the hooks through the AsyncWrap class. But there is nothing inherently correct
+in doing that, for example if v8 provided those facilities the AsyncWrap class
+would not need be involved in the AsyncWrap API._
 
-There was also some discussion in [issue #21](https://github.com/nodejs/tracing-wg/issues/21#issuecomment-142727693).
+For the remaining description the API part is what is meant by AsyncWrap.
+
+## Handle objects
+
+AsyncWrap emits events (hooks) that informs the consumer about the life of all handle
+objects in node. Thus in order to understand AsyncWrap one must first understand
+handle objects.
+
+The API that node exposes is mostly defined in JavaScript files. However
+ECMAScript does not define any API for creating a TCP socket, reading a file
```

Comment:
Maybe something like "Node's core API is mostly defined in JavaScript. However..."